### ISSUE TITLE
Add reference frames of planets and other bodies

### ISSUE THREAD

**Issue by Juanlu001** [astropy/astropy#5898]
Using the machinery to define new reference frames in Astropy and the information of this report of the  IAU Working Group on Cartographic Coordinates and Rotational Elements it should be possible to define the reference frames of the planets of the Solar System and other bodies:

https://astropedia.astrogeology.usgs.gov/download/Docs/WGCCRE/WGCCRE2009reprint.pdf

Is this something that would be in scope for Astropy? If not, I will try to implement them outside: https://github.com/poliastro/poliastro/issues/109

**Comment by Cadair** [comment]
I have looked into this report a little with the intent of defining a transformation between the SunPy solar frames.

Correct me if I am wrong, but the information in the report provides a reference at some point in time? To define a transformation into something like ICRS you would have to know how the reference points change over time?

**Comment by Juanlu001** [comment]
> Correct me if I am wrong, but the information in the report provides a reference at some point in time? To define a transformation into something like ICRS you would have to know how the reference points change over time?

Yes, that's how I understand it.

In the case of Mars, for example, we have [and now I miss LaTeX on GH issues]:

α_0 = 317.68143 - 0.1061 T
δ_0 = 52.88650 - 0.0609 T
W = 176.630 + 350.89198226 d

where (α_0, δ_0) are "ICRF equatorial coordinates at epoch J2000.0" of the north pole of the body, W is "the ephemeris position of the prime meridian", T is "interval in Julian centuries (of 36525 days) from the standard epoch" and d is "interval in days from the standard epoch".

![page-shot-2017-3-22 wgccre2009reprint pdf](https://cloud.githubusercontent.com/assets/316517/24196045/721765c2-0efc-11e7-9a3f-cd59a9e593f1.png)

**Comment by Cadair** [comment]
OK, so if we are to define these transformations, what can we use to calculate the offset at a given obs time?

**Comment by Juanlu001** [comment]
I'm not sure I understand, could you elaborate on what you're missing?

**Comment by Cadair** [comment]
I think I am not understanding it correctly. I am coming at this from the point of view of trying to convert from a coordinate frame that uses the Sun (centre and north pole) as a reference point.
If I want to go (for simplicity) from a point on the Sun today to ICRS I am going to need to compensate for the precession of the Sun from the epoch no?

**Comment by Juanlu001** [comment]
The coordinates of the poles of the bodies are referred to the ICRF, so if I understood correctly all the information is captured by the terms that depend on the passage of time. Notice that, for the case of the Sun, the evolution of the prime meridian is also considered:

α_0 = 286°.13
δ_0 = 63°.87
W = 84°.176 + 14°.1844000 d

**Comment by Cadair** [comment]
Ok, it's been a while since I looked at this, I will see if I can find the time to re-read the report and wrap my head around it again.

Thanks a lot for your help.

**Comment by mhvk** [comment]
@Juanlu001 - on your original question - my own sense is that this is in scope for astropy itself, partially because I feel it is important that astropy is the package to use if one wants to conform to IAU standards. Also, more practically, while for the Sun arguable `sunpy` is just as logical, for the planets I don't see another package (`poliastro` is not particularly logical in itself, though it is *very* logical that it would use the information if available).

In principle, given what you wrote, it should not be that difficult to implement in astropy either: we already have access to the planet ephemerides, so really it is a matter of defining appropriate reference frames (that rely not on `EarthLocation`, but something more general planet location; indeed, this may be a point where we generalize `EarthLocation` as a representation that can convert between lon, lat, height and cartesian; see https://github.com/astropy/astropy/issues/4261#issuecomment-150343826)

Anyway, it would be really cool if one could use astropy to tell where the planets would be in the sky as seen by a mars rover!!

**Comment by Juanlu001** [comment]
If it is in scope for Astropy itself then I might as well propose a PR. I wonder if it would be nice to agree on an API before start coding, what's the usual way to proceed? An APE, discussion on the mailing list, perhaps iterating on the wiki?

**Comment by pllim** [comment]
> it would be really cool if one could use astropy to tell where the planets would be in the sky as seen by a mars rover

And perhaps also a function call `find_matt_damon()`?

**Comment by mhvk** [comment]
I guess it would be good to lay out what would be needed, realizing there are two different purposes: knowing where a given location on the sun or a planet is (i.e., where it would be in one's telescope from earth), and using such a location to see where other things are. So, very broadly, I guess one would need
1. a generalized `EarthLocation`;
2. conversion of that to a `SkyCoord` so it can be localized;
3. use in sun/planet frames (in which it replaces `obsgeoloc`, `obsgeovel`).

Point 3 may require creating frames on demand (as for the offset frames).

My overall sense would be to start with the Sun, since then 1 and 2 may already be done in sunpy. I also feel that we could have those without worrying too much about generalizing to many planets. @cadair, what do you think?

**Comment by Cadair** [comment]
So in SunPy, the frames we have specify location of the observer in the heliosphere, with respect to the solar centre. So we specify location of the observers with an instance of one of our frames (currently a PR).

Things that would be useful to SunPy that I think this has the possibility to address or at least help with are:

* Position of the earth with respect to the solar equator.
* Location of the north pole of the Sun in ICRS coordinates (for defining a transformation from sunpy frames to ICRS)
* Calculation of the locations of planets in solar frames from arbitrary locations in the heliosphere (satellites).

**Comment by aelanman** [comment]
Has there been any further development on this?
I'm interested in using astropy to do ra/dec to az/alt coordinate transformations for observers located on the Moon.

**Comment by Cadair** [comment]
SunPy implemented the transformation to the solar frame using this reference which might be useful to anyone wanting to take a pass at this.

**Comment by Juanlu001** [comment]
I tried to implement frames parallel to ICRS centered in other bodies, but I was stopped by https://github.com/astropy/astropy/issues/7770.

**Comment by jarmokivekas** [comment]
I'd also use this feature if it were available. Specifically to get altitude and azimuth of the sun in reference to the local horizon on Mars or Moon.

The IAU Working Group on Cartographic Coordinates and Rotational Elements (WGCCRE) also maintains recommended datum ellipsoids/reference spheres for solar system planets required for proper handling of the lat, lon, height of the observer position. Although these should also be user-definable since people might use other datums and not all bodies have them. 

I may be willing to work on this if I don't find a solution from some other library.

**Comment by mhvk** [comment]
@jarmokivekas - help would definitely be appreciated!  And if you do find a solution elsewhere, please do post a link to it here -- we can perhaps use it too or copy ideas from it.

**Comment by jarmokivekas** [comment]
I did some digging and looks like the `EarthLocation` and `AltAz` classes and the associated `cirs_to_altaz` transform uses the ERFA/SOFA library python bindings for `astorpy._erfa.apio13(...)`, `astropy._erfa.atioq(...)`  on the back end to do the CIRS <-> Observed coordinate transfom, among other things.

I searched through the SOFA api reference, and as far as I'm aware, SOFA uses the WGS84 ellipsoid by default, and is therefore similarly Earth-centric, without obvious options on how to generalize to other planets, unfortunately. I've never used SOFA , and I didn't read any source code though, so there may be a workaround I don't know of.

Also, I've made a quick-and dirty python implementaion for Mars and Moon rotation and orientation models defined int the IAU WGCCRE reports (they are quite trivial). I'll see about pushing them to a branch at some point, just in case anybody is curious.

**Comment by mhvk** [comment]
For the ellipsoid, SOFA/ERFA can do conversion between XYZ and arbitrary ones: `eraGd2gce`, so at least that part should not be too difficult.

**Comment by jarmokivekas** [comment]
As a status update, I ended up not finding a ready-made solution, but used the `astropy` internals to implement this functionality myself using `get_body_barycentric` for planet position over time, and `CartesianRepresentation`, `ShpericalRepresentation`, and `rotation_matrix` etc. to implement the planet rotation and axis tilt transformations from the WGCCRE reports.

Here's a sample of a debug plot for my imaginary debug planet. The `topocentric direction` ray points toward the sun and provides the altitude and azimuth from the observer location. This solution is purely geometric for now. It doesn't take into account the travel time of light, atmospheric effects, or other fancy things. I will test it by putting it through the work I need it for and then see about a pull request if it actually works in a useful way. 
![image](https://user-images.githubusercontent.com/3074453/67830815-c8364c80-fae4-11e9-80e7-a55d13d32781.png)

I'm thinking of making an `EllipsoidalRepresentation` subclass of `astropy.coordinates.BaseRepresentation`, as well. The BaseRepresentation interface has worked well for me, but I found that this representation was missing. Probably a wrapper for some ERFA function.

**Comment by mhvk** [comment]
@jarmokivekas - nice!
And good idea about the ellipsoidal representation class. Indeed, we've been thinking about a similar approach for `EarthLocation` itself. See https://github.com/astropy/astropy/issues/4261#issuecomment-150343826. (The reason this was not done initially is that we needed `EarthLocation` for sidereal times in `astropy.time` before the coordinates machinery matured completely...). So, if you're up for it, that may be worth a separate PR. (It would perhaps be somewhat different from the regular representations in that it has parameters that control it).

**Comment by Juanlu001** [comment]
For those interested, we added fixed frames for planets in https://github.com/poliastro/poliastro/blob/master/src/poliastro/frames/fixed.py (still not super happy with the implementation but at least it's working)

**Comment by pllim** [comment]
@Juanlu001 , you forgot to take Pluto out. 😉

**Comment by jarmokivekas** [comment]
Despite hacking something together preliminary together, I will probably not end up doing any PRs for this. The poliastro implementation already covers much of the same things (and in a less hacky manner). Having topocentric (Alt, Az) frames co-rotating with the planet on top of that would sure be nice, though.

I ended up moving away from poliastro and astropy for my current project, and using [SPICE](https://naif.jpl.nasa.gov/naif/) with fixed-offset frame kernels instead. SPICE can be a little complicated to use, but it has a well-developed system for the topocentric coordinate frames I was looking for, so I'll stick with it for now.

**Comment by aelanman** [comment]
I have started working on some changes to address this, which you can find here [https://github.com/aelanman/astropy/tree/moon_locations](https://github.com/aelanman/astropy/tree/moon_locations). I started writing this before thinking to re-check this issue... This works by incorporating SPICE transformations directly in astropy via the spiceypy module.

There are two Moon-centered-moon-fixed (MCMF) reference frames in common use, described in this [white paper](https://lunar.gsfc.nasa.gov/library/LunCoordWhitePaper-10-08.pdf). I'm using the "mean Earth/Polar axis" frame (MOON_ME in SPICE), which sets the prime meridian of to the mean direction of the Earth and sets the North pole to the mean rotation axis. SPICE can make a rotation matrix to transform from ICRS (which it calls J2000) to the MOON_ME frame, but I'm not sure how to include stellar aberration from the Moon's motion.

The changes I've made define a `MoonLocation` class, which describes positions on the moon in the Moon_ME frame (cartesian "selenocentric" positions as well as "geodetic" lat/lon/height, where height is with respect to a sphere of fixed radius). I'm not sure how SPICE kernels describe the shape of the moon mathematically. Is there a more precise ellipsoid model?

**Comment by jarmokivekas** [comment]
@aelanman SPICE is a complex system, so there are many ways to represent things like the moon surface, but at least the spheroids/ellipsoids defined in the "IAU2000" spatial reference recommendations are readily available.  I would recommend using those.

I usually look up IAU recommended ellipsoids here:
https://spatialreference.org/ref/iau2000/?search=IAU+2000&srtext=Search

For example  `IAU2000:30100` "Moon 2000" ellipsoid is `+proj=longlat +a=1737400 +b=1737400 +no_defs` in proj4 notation.

**Comment by aelanman** [comment]
@jarmokivekas Thanks for the suggestion! From your comment, it sounds like the reference ellipsoid for the Moon is a sphere anyway?

Since my last comment, I've gone in a slightly different direction with this. I've moved my changes to a new package ([github.com/aelanman/lunarsky](https://github.com/aelanman/lunarsky)), which includes the new frames and a modification to `SkyCoord` as well as tests. If this works out well, I would like to make the changes to astropy instead of having a separate module, but for now I can work with this.

**Comment by cmarmo** [comment]
Hi @astrojuanlu and all people working about this issue. Thanks a lot for trying to address the insertion of Solar System bodies reference systems in astropy! The entire planetary community is struggling with interoperability issues, their solution might depend on this implementation. I've never dived into the astropy code so this will probably take me some time... but may I ask if adding some more planetary related keywords to the WCS as from the [work done](https://agupubs.onlinelibrary.wiley.com/doi/10.1029/2018EA000388) during the [EuroplanetH2020](http://www.europlanet-2020-ri.eu/) could helpful in some way?

**Comment by mhvk** [comment]
@cmarmo - I'll let @astrojuanlu answer more about the details, but it does seem logical that `astropy.io.fits` would be able to interpret any planetary keywords and `astropy.wcs` represent the corresponding coordinate systems. It would seem that would get us some way to defining the coordinate systems themselves too! 

And at least from the outside, this looks exactly like the step-by-step approach that would be easiest to implement (and where every step is useful on its own). Before diving in any code, it might be good to open a first issue specifically about supporting the relevant FITS keywords and what would be needed in WCS, giving some explicit details.

One good thing is that @astrojuanlu has just been adding generalizations of the geodetic representations in #11086, which I think would form a good framework to build a general ellipsoid model on for the `[ABC]_RADIUS` mapping in the reference you gave. The idea is that we'll use the geodetic representations in `EarthLocation` (instead of it just being the representation), making it more like a coordinate frame object. This should make it much easier to have equivalent `PlanetLocations`...

**Comment by cmarmo** [comment]
Thanks @mhvk for your answer.

> Before diving in any code, it might be good to open a first issue specifically about supporting the relevant FITS keywords and what would be needed in WCS, giving some explicit details.

I will do that ASAP.

**Comment by astrojuanlu** [comment]
Hi @cmarmo , thanks for drawing my attention (again!) to this issue. It's been a while since I last worked on it, but according to my old comments, this is soft blocked on https://github.com/astropy/astropy/issues/7770 - which shouldn't be difficult to fix anyway. I have a few other `astropy.coordinates` tasks I'd like to work on, but I'm definitely raising the priority of this one.

**Comment by pllim** [comment]
To those still interested in this issue, please see https://github.com/astropy/astropy/issues/11170 for proposed path forward. Thanks!

**Comment by jacopok** [comment]
Hello, I came across this issue while looking for the functionality of transforming between an ICRS solar-centered frame and one centered on the Moon, with axes rotating so that they remain aligned to its surface. 
It seems like this is completely implemented in @aelanman 's package [lunarsky](https://github.com/aelanman/lunarsky). 

I tried to read #11170 and cannot really understand what's going on - will it implement equivalent functionality? Has it already done so? It seems like it's already closed and the remaining issues are minor, but I cannot find the functionality (documented) in the current version of astropy, 5.3.4.

My main question is: for a new project requiring this functionality, should we use @aelanman 's package or will there be / is there already some native astropy implementation?

**Comment by cmarmo** [comment]
Hi @jacopok the work in #11170 will be available in Astropy 6.0. I have summed up the implementation on the [OpenPlanetary forum](https://openplanetary.discourse.group/t/planetary-body-fixed-reference-frame-wcs-description-in-astropy-6-0/719) if you want to have a look.
Spheroidal body-fixed reference frames only are implemented in 6.0: transformations between, let's say, ICRS and body-fixd are not available yet.

**Comment by jacopok** [comment]
I see, thanks! So there is no current plan to implement the transformations as well, correct?
Is that within the scope of this issue or should an independent one be made to track that feature request?

**Comment by cmarmo** [comment]
> I see, thanks! So there is no current plan to implement the transformations as well, correct?

Let's say there is the intention of , but not current plan to : it needs some time and funding...

> Is that within the scope of this issue or should an independent one be made to track that feature request?

I believe this is within the scope of the issue, but I let the maintainers comment on this.

**Comment by mhvk** [comment]
I think indeed the present issue was to ask whether it is possible to transform coordinates to apparent ones on any solar system body. I think astropy should certainly provide the ability to do so - for which we now have a very nice start with @cmarmo's work. I'm less sure whether the actual planetary constants belong; they might more logically be in a specific solar system package (like sunpy provides for the sun). (At least, sadly it does not seem likely a typical astronomer will soon analyze data observed from other planets than Earth...). But in the end, astropy is driven forward by people who make PRs!

**Comment by Caraffa-git** [comment]
> If this works out well, I would like to make the changes to astropy instead of having a separate module, but for now I can work with this.

@aelanman did you made any progress in this direction? I've come across similar problem, where i have to determine distance between defined point on the moon and specific location on the earth. I don't know if there is a way to do it with the astropy alone but it's kinda off-topic. (I'm not an astrophysicist)

**Comment by aelanman** [comment]
Hi @Caraffa-git -- As far I'm aware, the last astropy development on that is @cmarmo 's planetary WCS work, but I haven't been following very closely.

My package lunarsky can certainly do what you're asking, in conjunction with astropy. Feel free to contact me if you want any help with it! You would create a `MoonLocation` instance with the position you want there, and an `EarthLocation` instance with the Earth location. Then you'd transform them to a common reference frame (e.g., ITRS for a geocentric position) and measure the cartesian distance between them. Whether or not this is close enough to correct will depend on your application.
